walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,146 @@
+    int meetingDuration) {
+    List<TimeRange> validTimeRanges = new ArrayList<>();
+
+    // Get the inverse TimeRanges of the invalidTimeRanges.
+    int index = 0;
+    int start = TimeRange.START_OF_DAY;
+    int end = TimeRange.START_OF_DAY;
+    while (index < invalidTimeRanges.size()) {
+      end = invalidTimeRanges.get(index).start();
+
+      // Add new valid TimeRange to list.
+      if (start != end && fitsMeetingDuration(end - start, meetingDuration)) {
+        validTimeRanges.add(TimeRange.fromStartEnd(start, end, false));
+      }
+
+      // Get the start of the next valid TimeRange, and increment index.
+      start = invalidTimeRanges.get(index).end();
+      index++;
+    }
+
+    // Get the last valid TimeRange, if applicable.
+    if (start != TimeRange.END_OF_DAY && 
+      fitsMeetingDuration(TimeRange.END_OF_DAY - start, meetingDuration)) {
+
+      validTimeRanges.add(
+        TimeRange.fromStartEnd(start, TimeRange.END_OF_DAY, true));
+    }
+
+    return validTimeRanges;
+  }
+
+  /**
+   * Merge any conflicting TimeRanges in the unmergedTimeRanges list.
+   * The unmergedTimeRanges list must be sorted by start-time. This will modify
+   * the original list, so no return is necessary.
+   */
+  private void mergeConflictingTimeRanges(List<TimeRange> unmergedTimeRanges) {
+      
+    int index = 0;
+    while (index < unmergedTimeRanges.size() - 1) {
+      TimeRange firstTimeRange = unmergedTimeRanges.get(index);
+      TimeRange secondTimeRange = unmergedTimeRanges.get(index + 1);
+
+      // Check if TimeRanges conflict, and if so, merge them.
+      if (firstTimeRange.overlaps(secondTimeRange)) {
+        // Create merged TimeRange.
+        int endMergedTime = (firstTimeRange.end() > secondTimeRange.end()) ?
+          firstTimeRange.end() : secondTimeRange.end();
+        TimeRange mergedTimeRange = TimeRange.fromStartEnd(firstTimeRange.start(), 
+          endMergedTime, false);
+
+        // Remove old TimeRanges, and add new merged TimeRange.
+        unmergedTimeRanges.remove(firstTimeRange);
+        unmergedTimeRanges.remove(secondTimeRange);
+        unmergedTimeRanges.add(index, mergedTimeRange);
+      } else {
+        index++;
+      }
+    }
+  }
+
+  /**
+   * Return all of the invalid TimeRanges (conflicting events) to a list.
+   * This method modifies the original list, so no return is necessary.
+   */
+  private void populateInvalidTimeRanges(List<TimeRange> invalidTimeRanges, 
+    Collection<Event> events, Collection<String> attendees) {
+
+    // Add event TimeRange if event has attendees in current meeting query.
+    for (Event e : events) {
+      if (containsElement(attendees, e.getAttendees())) {
+        invalidTimeRanges.add(e.getWhen());
+      }
+    }
+  }
+
+  /**
+   * Return true if l1 contains any element in l2; otherwise, false.
+   */
+  private boolean containsElement(Collection<String> l1, Collection<String> l2) {
```

Comment:
food for thought. I don't know if this method is really necessary. If you kept this code in the populateInvalidTimeRanges method it would look like:

for (Event e : events) {
   for (String attendee : e.getAttendees()) {
      if (attendees.contains(attendee)) {
        invalidTimeRanges.add(e.getWhen());
        break;
      }
    }

That takes fewer lines, and isn't difficult to understand. Breaking out a section of code to avoid breaks and increase readability is a great strategy, but sometimes it's a bit unnecessary. Does that make sense?